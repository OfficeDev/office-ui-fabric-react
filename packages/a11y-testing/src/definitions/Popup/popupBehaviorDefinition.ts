import { Rule } from './../../types';
import { BehaviorRule } from './../../rules/rules';

export const popupBehaviorDefinitionTriggerSlotNotTabbable: Rule[] = [
  BehaviorRule.slot('trigger')
    .forProps({ tabbableTrigger: true })
    .hasAttribute('tabIndex', '0')
    .hasAttribute('aria-haspopup', 'dialog')
    .description(`if 'tabbableTrigger' is 'true'.`),
  BehaviorRule.slot('trigger')
    .forProps({ tabbableTrigger: true })
    .hasAttribute('tabIndex', '0')
    .description(`if it is 'not tabbable' and 'tabbableTrigger' is 'true'.`),
  BehaviorRule.slot('trigger')
    .forProps({ tabbableTrigger: false })
    .doesNotHaveAttribute('tabIndex')
    .description(`if 'tabbableTrigger' is 'false'.`),
  BehaviorRule.slot('trigger')
    .forProps({ tabbableTrigger: false })
    .doesNotHaveAttribute('aria-haspopup')
    .description(`if 'tabbableTrigger' is 'false'.`),
];

export const popupBehaviorDefinitionTriggerSlotTabbable: Rule[] = [
  BehaviorRule.slot('trigger')
    .forProps({ tabbableTrigger: true })
    .hasAttribute('aria-haspopup', 'dialog')
    .description(`if 'tabbableTrigger' is 'true'.`),
  BehaviorRule.slot('trigger')
    .forProps({ tabbableTrigger: true })
    .doesNotHaveAttribute('tabIndex')
    .description(`if it is 'tabbable' and 'tabbableTrigger' is 'true'.`),
];

export const popupBehaviorDefinitionTriggerSlotWithTabIndex: Rule[] = [
  BehaviorRule.slot('trigger')
    .forProps({ tabbableTrigger: true })
    .hasAttribute('tabIndex', '-1')
    .description(`if current 'trigger' element has 'tabindex=-1' and 'tabbableTrigger' is 'true'.`),
];

export const popupBehaviorDefinitionPopupSlot: Rule[] = [
  BehaviorRule.slot('popup')
    .forProps({ trapFocus: true })
    .hasAttribute('role', 'dialog')
    .description(`if 'trapFocus' is 'true'.`),
  BehaviorRule.slot('popup')
    .forProps({ trapFocus: true })
    .hasAttribute('aria-modal', 'true')
    .description(`if 'trapFocus' is 'true'.`),
  BehaviorRule.slot('popup')
    .forProps({ trapFocus: false })
    .hasAttribute('role', 'complementary')
    .description(`if 'trapFocus' is 'false'.`),
  BehaviorRule.slot('popup')
    .forProps({ trapFocus: false })
    .doesNotHaveAttribute('aria-modal')
    .description(`if 'trapFocus' is 'false'.`),
];

// remove 'definition' which contains 'dialog' rule because description is already cover in popupBehaviorDefinitionTriggerSlotNotTabbable
// just want to keep 'definition' in the array, in order to test 'tabbable' and 'notTabbable' trigger
const modifiedForDocumentation = popupBehaviorDefinitionTriggerSlotTabbable.filter(
  definition => definition.stringify().search('dialog') === -1,
);

export const popupBehaviorDefinition = popupBehaviorDefinitionTriggerSlotNotTabbable
  .concat(modifiedForDocumentation)
  .concat(popupBehaviorDefinitionTriggerSlotWithTabIndex)
  .concat(popupBehaviorDefinitionPopupSlot);
