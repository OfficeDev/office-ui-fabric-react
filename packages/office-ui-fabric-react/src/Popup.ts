export * from './packages/react-surfaces/Popup';
