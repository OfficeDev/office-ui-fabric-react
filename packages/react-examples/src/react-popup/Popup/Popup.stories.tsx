import * as React from 'react';
import { Popup, PopupTrigger, PopupContent, PopupProps } from '@fluentui/react-popup';
import { makeStyles } from '@fluentui/react-make-styles';

const useStyles = makeStyles({
  example: {
    padding: '12px',
  },
});

const ExampleContent = () => {
  const { example } = useStyles();
  return (
    <div className={example}>
      <h3>Popup content</h3>

      <div>This is some popup content</div>
    </div>
  );
};

export const Default = (props: PopupProps) => (
  <Popup {...props}>
    <PopupTrigger>
      <button>Popup trigger</button>
    </PopupTrigger>

    <PopupContent>
      <ExampleContent />
    </PopupContent>
  </Popup>
);

Default.argTypes = {
  openOnContext: {
    defaultValue: false,
    control: 'boolean',
  },

  openOnHover: {
    defaultValue: false,
    control: 'boolean',
  },
};

export const AnchorToTarget = () => {
  const [target, setTarget] = React.useState<HTMLButtonElement | null>();

  return (
    <>
      <div>
        <Popup target={target}>
          <PopupTrigger>
            <button>Popup trigger</button>
          </PopupTrigger>

          <PopupContent>
            <ExampleContent />
          </PopupContent>
        </Popup>
      </div>

      <button style={{ marginTop: 100 }} ref={setTarget}>
        Custom target
      </button>
    </>
  );
};

export const Controlled = () => {
  const [open, setOpen] = React.useState(false);
  const onOpenChange: PopupProps['onOpenChange'] = (_, data) => setOpen(data.open || false);

  return (
    <>
      <Popup open={open} onOpenChange={onOpenChange}>
        <PopupTrigger>
          <button>Custom trigger</button>
        </PopupTrigger>
        <PopupContent>
          <ExampleContent />
        </PopupContent>
      </Popup>
    </>
  );
};

export const WithCustomTrigger = () => {
  const [open, setOpen] = React.useState(false);
  const [target, setTarget] = React.useState<HTMLElement | null>(null);
  const onClick = () => setOpen(s => !s);
  const onKeyDown = (e: React.KeyboardEvent) => {
    if (e.key === 'Enter' || e.key === ' ') {
      setOpen(s => !s);
    }
  };
  const onOpenChange: PopupProps['onOpenChange'] = (_, data) => setOpen(data.open || false);

  return (
    <>
      <div
        tabIndex={0}
        style={{ width: 100, border: '1px solid red' }}
        aria-haspopup
        ref={setTarget}
        onClick={onClick}
        onKeyDown={onKeyDown}
      >
        Custom trigger
      </div>
      <Popup target={target} open={open} onOpenChange={onOpenChange}>
        <PopupContent>
          <ExampleContent />
        </PopupContent>
      </Popup>
    </>
  );
};
