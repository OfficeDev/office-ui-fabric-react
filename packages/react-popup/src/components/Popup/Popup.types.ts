import * as React from 'react';
import { PopperOptions } from '@fluentui/react-positioning';
import { PortalProps } from '@fluentui/react-portal';

/**
 * Popup Props
 */
export interface PopupProps
  extends Pick<PopperOptions, 'position' | 'align' | 'offset'>,
    Pick<PortalProps, 'mountNode'> {
  children: React.ReactNode;
  /**
   * Controls the opening of the popup
   */
  open?: boolean;
  /**
   * Used to set the initial open state of the popup in uncontrolled mode
   */
  defaultOpen?: boolean;
  /**
   * Uses a custom target HTML element to anchor the popup
   */
  target?: HTMLElement | null;
  /**
   * Call back when the component requests to change value
   * The `open` value is used as a hint when directly controlling the component
   */
  onOpenChange?: (e: OpenPopupEvents, data: OnOpenChangeData) => void;
}

/**
 * Names of the shorthand properties in PopupProps
 */
export const popupShorthandProps = [] as const;

/**
 * Names of PopupProps that have a default value in usePopup
 */
export type PopupDefaultedProps = never;

/**
 * Popup State
 */
export interface PopupState extends PopupProps {
  /**
   * Open state of the popup
   */
  open: boolean;
  /**
   * Callback to open/close the popup
   */
  setOpen: (e: OpenPopupEvents, open: boolean) => void;
  /**
   * Ref of the PopupTrigger
   */
  triggerRef: React.MutableRefObject<HTMLElement | null>;
  /**
   * Ref of the PopupContent
   */
  contentRef: React.MutableRefObject<HTMLElement | null>;
  /**
   * Flag to open the popup by hovering the trigger
   */
  openOnHover?: boolean;
  /**
   * Flag to open the popup as a context menu. Disables all other interactions
   */
  openOnContext?: boolean;
}

/**
 * Data attached to open/close events
 */
export interface OnOpenChangeData extends Pick<PopupState, 'open'> {}

/**
 * The supported events that will trigger open/close of the menu
 */
export type OpenPopupEvents =
  | MouseEvent
  | TouchEvent
  | React.MouseEvent<HTMLElement>
  | React.KeyboardEvent<HTMLElement>
  | React.FocusEvent<HTMLElement>;
