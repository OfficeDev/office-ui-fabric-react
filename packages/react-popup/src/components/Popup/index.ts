export * from './Popup';
export * from './Popup.types';
export * from './renderPopup';
export * from './usePopup';
