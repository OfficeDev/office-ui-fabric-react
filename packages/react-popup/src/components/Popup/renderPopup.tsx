import * as React from 'react';
import { PopupContext } from '../../popupContext';
import { PopupState } from './Popup.types';

/**
 * Render the final JSX of Popup
 */
export const renderPopup = (state: PopupState) => {
  const { open, setOpen, triggerRef, contentRef, target, openOnContext, openOnHover, mountNode } = state;

  return (
    <PopupContext.Provider
      value={{ open, setOpen, triggerRef, contentRef, target, openOnHover, openOnContext, mountNode }}
    >
      {state.children}
    </PopupContext.Provider>
  );
};
