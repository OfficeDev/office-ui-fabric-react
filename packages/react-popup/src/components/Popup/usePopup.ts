import * as React from 'react';
import { makeMergeProps, useControllableValue, useOnClickOutside, useEventCallback } from '@fluentui/react-utilities';
import { useFluent } from '@fluentui/react-shared-contexts';
import { usePopper } from '@fluentui/react-positioning';
import { PopupProps, PopupState } from './Popup.types';
import { useCallbackRef } from './useCallbackRef';

const mergeProps = makeMergeProps<PopupState>({});

/**
 * Create the state required to render Popup.
 *
 * The returned state can be modified with hooks such as usePopupStyles,
 * before being passed to renderPopup.
 *
 * @param props - props from this instance of Popup
 * @param defaultProps - (optional) default prop values provided by the implementing type
 */
export const usePopup = (props: PopupProps, defaultProps?: PopupProps): PopupState => {
  const state = mergeProps(
    {
      open: (undefined as unknown) as boolean, // mergeProps typings require this
      setOpen: () => null,
      triggerRef: { current: null },
      targetRef: { current: null },
      contentRef: { current: null },
      children: null,
      position: 'below',
      align: 'start',
    },
    defaultProps,
    props,
  );

  useOpenState(state);
  usePopupRefs(state);

  const { targetDocument } = useFluent();
  useOnClickOutside({
    element: targetDocument,
    callback: ev => state.setOpen(ev, false),
    refs: [state.triggerRef, state.contentRef, state.targetRef],
  });

  return state;
};

/**
 * Creates and manages the popup open state
 * @param state Popup state
 */
function useOpenState(state: PopupState): PopupState {
  const [open, setOpen] = useControllableValue(state.open, state.defaultOpen);
  // TODO fix useControllableValue typing
  state.open = open !== undefined ? open : state.open;
  const onOpenChange: PopupState['onOpenChange'] = useEventCallback((e, data) => state.onOpenChange?.(e, data));
  state.setOpen = React.useCallback(
    (e, shouldOpen) => {
      setOpen(prevOpen => {
        // More than one event (mouse, focus, keyboard) can request the popup to close
        // We assume the first event is the correct one
        if (prevOpen !== shouldOpen) {
          onOpenChange?.(e, { open: shouldOpen });
        }

        return shouldOpen;
      });
    },
    [setOpen, onOpenChange],
  );

  return state;
}

/**
 * Creates and sets the necessary trigger, target and content refs used by popup
 * @param state Popup state
 */
function usePopupRefs(state: PopupState): PopupState {
  const { targetRef, containerRef: contentRef } = usePopper({
    align: state.align,
    position: state.position,
  });

  state.contentRef = contentRef;
  state.targetRef = targetRef;
  if (state.target) {
    state.targetRef.current = state.target;
  }

  const triggerRef = useCallbackRef<HTMLElement>(null, (newValue, lastValue) => {
    if (newValue !== lastValue && !state.target) {
      state.targetRef.current = (newValue as unknown) as HTMLElement;
    }
  });
  state.triggerRef = triggerRef;

  return state;
}
