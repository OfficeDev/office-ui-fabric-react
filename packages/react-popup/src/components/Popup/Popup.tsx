import * as React from 'react';
import { usePopup } from './usePopup';
import { PopupProps } from './Popup.types';
import { renderPopup } from './renderPopup';

/**
 * Popup component
 */
export const Popup: React.FC<PopupProps> = props => {
  const state = usePopup(props);

  return renderPopup(state);
};

Popup.displayName = 'Popup';
