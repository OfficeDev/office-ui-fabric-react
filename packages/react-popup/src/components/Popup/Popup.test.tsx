import * as React from 'react';
import { Popup } from './Popup';
import * as renderer from 'react-test-renderer';
import { ReactWrapper } from 'enzyme';

describe('Popup', () => {
  let wrapper: ReactWrapper | undefined;

  afterEach(() => {
    if (wrapper) {
      wrapper.unmount();
      wrapper = undefined;
    }
  });

  /**
   * Note: see more visual regression tests for Popup in /apps/vr-tests.
   */
  it('renders a default state', () => {
    const component = renderer.create(<Popup>Default Popup</Popup>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
