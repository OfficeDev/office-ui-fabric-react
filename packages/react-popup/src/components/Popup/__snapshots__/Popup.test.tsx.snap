// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popup renders a default state 1`] = `"Default Popup"`;
